File: kgk_customisations/kgk_customisations/report/f.m_process_performance_report/f.m_process_performance_report.js
// Copyright (c) 2025, Apjakal IT Solutions and contributors
// For license information, please see license.txt

frappe.query_reports["F.M Process Performance Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start(),
			"reqd": 1
		},
		{
			"fieldname": "to_date", 
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_end(),
			"reqd": 1
		},
		{
			"fieldname": "department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Section"
		},
		{
			"fieldname": "factory_process",
			"label": __("Factory Process"),
			"fieldtype": "Link",
			"options": "Factory Process"
		},
		{
			"fieldname": "day_type",
			"label": __("Day Type"),
			"fieldtype": "Select",
			"options": "All\nNormal\nWeekend",
			"default": "All"
		}
	],
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Color formatting for difference values
		if (column.fieldname === "diff" || column.fieldname === "diff_percentage") {
			if (flt(data[column.fieldname]) < 0) {
				value = `<span style="color: #d73527; font-weight: bold;">${value}</span>`;
			} else if (flt(data[column.fieldname]) > 0) {
				value = `<span style="color: #5cb85c; font-weight: bold;">${value}</span>`;
			}
		}
		
		// Highlight factory process names
		if (column.fieldname === "factory_process") {
			value = `<strong>${value}</strong>`;
		}
		
		return value;
	},
	"onload": function(report) {
		// Add custom CSS for better visualization
		if (!$('#process-performance-styles').length) {
			$('<style id="process-performance-styles">')
				.text(`
					.dt-row .dt-cell[data-col-index="1"],
					.dt-row .dt-cell[data-col-index="2"],
					.dt-row .dt-cell[data-col-index="3"] {
						text-align: right;
					}
					.negative-diff {
						background-color: #ffe6e6 !important;
					}
					.positive-diff {
						background-color: #e6ffe6 !important;
					}
				`)
				.appendTo('head');
		}
		
		// Auto-refresh functionality for real-time updates
		setInterval(function() {
			if (report.page.current_route && report.page.current_route[0] === 'query-report' 
				&& report.page.current_route[1] === 'F.M Process Performance Report') {
				report.refresh();
			}
		}, 300000); // 5 minutes
	},
	"get_datatable_options": function(options) {
		return Object.assign(options, {
			checkboxColumn: false,
			layout: 'fluid',
			rowHeight: 30,
			headerHeight: 35,
			cellHeight: 30,
			getRowClass: function(row) {
				if (row && row[3] < 0) { // diff column
					return 'negative-diff';
				} else if (row && row[3] > 0) {
					return 'positive-diff';
				}
				return '';
			}
		});
	}
};

frappe.query_reports["F.M Process Performance Report"] = {
	"filters": [

	]
};

File: kgk_customisations/kgk_customisations/report/f.m_process_performance_report/f.m_process_performance_report.py
# Copyright (c) 2025, Apjakal IT Solutions and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate
import math

def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	chart = get_chart_data(data, filters)
	summary = get_summary(data)
	return columns, data, None, chart, summary

def get_columns():
	return [
		{
			"fieldname": "factory_process",
			"fieldtype": "Link",
			"label": _("Factory Process"),
			"options": "Factory Process",
			"width": 300,
			"filterable": 0,
			"align": "center"
		},
		{
			"fieldname": "target",
			"fieldtype": "Int",
			"label": _("Target"),
			"width": 200,
			"filterable": 0,
			"align": "center"
		},
		{
			"fieldname": "actual",
			"fieldtype": "Int",
			"label": _("Actual"),
			"width": 200,
			"filterable": 0,
			"align": "center"
		},
		{
			"fieldname": "diff",
			"fieldtype": "Int",
			"label": _("Diff"),
			"width": 200,
			"filterable": 0,
			"align": "center"
		},
		{
			"fieldname": "diff_percentage",
			"fieldtype": "Int",
			"label": _("Diff %"),
			"width": 200,
			"filterable": 0,
			"align": "center"
		}
	]

def get_data(filters):
	if not filters:
		return []
		
	conditions = get_conditions(filters)
	
	# Get data from Factory Main and Factory Main Item
	data = frappe.db.sql(f"""
		SELECT 
			fmi.type as factory_process,
			SUM(CAST(fmi.target AS DECIMAL(10,2))) as target,
			SUM(fmi.actual) as actual
		FROM 
			`tabFactory Main` fm
		INNER JOIN 
			`tabFactory Main Item` fmi ON fm.name = fmi.parent
		WHERE 
			fm.docstatus < 2
			AND fmi.type IS NOT NULL 
			AND fmi.type != ''
			{conditions}
		GROUP BY 
			fmi.type
		ORDER BY 
			fmi.type
	""", filters, as_dict=1)
	
	# Calculate differences and percentages
	for row in data:
		target = math.ceil(flt(row.target))
		actual = math.ceil(flt(row.actual))
		row.target = target
		row.actual = actual
		row.diff = actual - target
		row.diff_percentage = math.ceil((row.diff / target * 100)) if target else 0
	
	return data

def get_conditions(filters):
	conditions = []
	
	if filters.get("from_date"):
		conditions.append("fm.work_date >= %(from_date)s")
	
	if filters.get("to_date"):
		conditions.append("fm.work_date <= %(to_date)s")
		
	if filters.get("department"):
		conditions.append("fm.section = %(department)s")
		
	if filters.get("factory_process"):
		conditions.append("fmi.type = %(factory_process)s")
		
	if filters.get("day_type") and filters.get("day_type") != "All":
		if filters.get("day_type") == "Normal":
			conditions.append("DAYOFWEEK(fm.work_date) NOT IN (1, 7)")  # Not Sunday or Saturday
		elif filters.get("day_type") == "Weekend":
			conditions.append("DAYOFWEEK(fm.work_date) IN (1, 7)")  # Sunday or Saturday
	
	return " AND " + " AND ".join(conditions) if conditions else ""

def get_chart_data(data, filters):
	if not data:
		return None
	
	processes = []
	targets = []
	actuals = []
	
	for row in data:
		processes.append(row.factory_process or "Unknown")
		targets.append(math.ceil(flt(row.target)))
		actuals.append(math.ceil(flt(row.actual)))
	
	return {
		"data": {
			"labels": processes,
			"datasets": [
				{
					"name": "Target",
					"chartType": "bar",
					"values": targets,
					"color": "#FF6B6B"
				},
				{
					"name": "Actual",
					"chartType": "bar", 
					"values": actuals,
					"color": "#4ECDC4"
				}
			]
		},
		"type": "bar",
		"height": 300,
		"colors": ["#FF6B6B", "#4ECDC4"]
	}

def get_summary(data):
	if not data:
		return []
	
	total_target = math.ceil(sum(flt(row["target"]) for row in data))
	total_actual = math.ceil(sum(flt(row["actual"]) for row in data))
	total_diff = total_actual - total_target
	total_diff_percentage = math.ceil((total_diff / total_target * 100)) if total_target else 0
	
	# Count processes with positive and negative variance
	positive_processes = len([row for row in data if flt(row["diff"]) > 0])
	negative_processes = len([row for row in data if flt(row["diff"]) < 0])
	
	return [
		{
			"value": len(data),
			"label": "Total Processes",
			"datatype": "Int",
			"currency": None
		},
		{
			"value": total_target,
			"label": "Total Target",
			"datatype": "Int",
			"currency": None
		},
		{
			"value": total_actual,
			"label": "Total Actual",
			"datatype": "Int",
			"currency": None
		},
		{
			"value": total_diff,
			"label": "Overall Difference",
			"datatype": "Int",
			"currency": None,
			"indicator": "Green" if total_diff >= 0 else "Red"
		},
		{
			"value": total_diff_percentage,
			"label": "Overall Performance",
			"datatype": "Int",
			"currency": None,
			"indicator": "Green" if total_diff_percentage >= 0 else "Red"
		},
		{
			"value": positive_processes,
			"label": "Above Target",
			"datatype": "Int",
			"currency": None,
			"indicator": "Green"
		},
		{
			"value": negative_processes,
			"label": "Below Target",
			"datatype": "Int",
			"currency": None,
			"indicator": "Red"
		}
	]
